File: frontend/src/routes/room/[room]/message.ts
enum MessageType {
	JOIN_ROOM = 'join_room',
	USER_JOINED = 'user_joined',
	SUBMIT_ESTIMATE = 'submit_estimate',
	REVEAL_ESTIMATE = 'reveal_estimate',
	ESTIMATE_REVEALED = 'estimate_revealed',
	RESET_PLANNING = 'reset_planning',
	PLANNING_RESETED = 'planning_reseted',
	USER_DISCONNECTED = 'user_disconnected',
	CONFIRM_CONNECTION = 'confirm_connection',
	CONFIRM_ESTIMATE_SUBMISSION = 'confirm_estimate_submission',
	USER_SUBMITTED_ESTIMATE = 'user_submitted_estimate'
}

interface MessageInterface {
	type: MessageType;
}

interface UserDTO {
	userName: string;
	uuid: string;
	estimate: string;
}

interface RoomDTO {
	roomCode: string;
	roomName: string;
	isEstimateRevealed: boolean;
}

// sended

interface JoinRoomMessage extends MessageInterface {
	type: MessageType.JOIN_ROOM;
	payload: {
		roomCode: string;
		nickname: string;
	};
}

interface SubmitEstimateMessage {
	type: MessageType.SUBMIT_ESTIMATE;
	payload: {
		estimate: string;
	};
}

interface RevealEstimateMessage {
	type: MessageType.REVEAL_ESTIMATE;
	payload: {
		shouldReveal: boolean;
	};
}

interface ResetPlanningMessage {
	type: MessageType.RESET_PLANNING;
}

// received

interface ConfirmConnectionMessage {
	type: MessageType.CONFIRM_CONNECTION;
	payload: {
		user: UserDTO;
		connectedUsers: UserDTO[];
		room: RoomDTO;
	};
}

interface UserDisconnectedMessage {
	type: MessageType.USER_DISCONNECTED;
	payload: {
		user: UserDTO;
	};
}

interface UserJoinedMessage {
	type: MessageType.USER_JOINED;
	payload: {
		user: UserDTO;
	};
}

interface ConfirmEstimateSubmittedMessage {
	type: MessageType.CONFIRM_ESTIMATE_SUBMISSION;
	payload: {
		estimate: string;
	};
}

interface UserSubmittedEstimateMessage {
	type: MessageType.USER_SUBMITTED_ESTIMATE;
	payload: {
		estimate: string;
		user: UserDTO;
	};
}

interface EstimateRevealedMessage {
	type: MessageType.ESTIMATE_REVEALED;
	payload: {
		estimates: Record<string, number>;
	};
}

interface PlanningResetedMessage {
	type: MessageType.PLANNING_RESETED;
}

// Define a union type for all possible message types
type Message =
	| JoinRoomMessage
	| UserJoinedMessage
	| SubmitEstimateMessage
	| EstimateRevealedMessage
	| ResetPlanningMessage
	| UserDisconnectedMessage
	| ConfirmConnectionMessage
	| ConfirmEstimateSubmittedMessage
	| UserSubmittedEstimateMessage
	| RevealEstimateMessage
	| PlanningResetedMessage;

export { MessageType };
export type { JoinRoomMessage, Message, SubmitEstimateMessage, RevealEstimateMessage, ResetPlanningMessage };

//TODO: refactor/clean this file

File: server/message.go
package main

import (
	"encoding/json"

	rm "github.com/baptiste-requet/plan-poker-go/rooms-manager"
)

type User struct {
	UserName string `json:"userName"`
	Uuid     string `json:"uuid"`
	Estimate string `json:"estimate"`
}

type Room struct {
	RoomCode           string `json:"roomCode"`
	RoomName           string `json:"roomName"`
	IsEstimateRevealed bool   `json:"isEstimateRevealed"`
}

type MessageType string

/*
	Both received and send message
*/

const (
	REVEAL_ESTIMATE MessageType = "reveal_estimate"
)

type RevealEstimateMessage struct {
	ShouldReveal bool `json:"shouldReveal"`
}

/*
	Received messages
*/

const (
	JOIN_ROOM       MessageType = "join_room"
	SUBMIT_ESTIMATE MessageType = "submit_estimate"
	RESET_PLANNING  MessageType = "reset_planning"
)

type ReceiveMessage struct {
	Type    MessageType
	Payload json.RawMessage
}

type JoinRoomMessage struct {
	RoomCode string `json:"roomCode"`
	Nickname string `json:"nickname"`
}

type SubmitEstimateMessage struct {
	Estimate string `json:"estimate"`
}

type ResetPlanningMessage struct {
}

/*
	Sended messages
*/

const (
	USER_JOINED                 MessageType = "user_joined"
	USER_DISCONNECTED           MessageType = "user_disconnected"
	CONFIRM_CONNECTION          MessageType = "confirm_connection"
	CONFIRM_ESTIMATE_SUBMISSION MessageType = "confirm_estimate_submission"
	USER_SUBMITTED_ESTIMATE     MessageType = "user_submitted_estimate"
	ESTIMATE_SUBMITTED          MessageType = "estimate_submitted"
	ESTIMATE_REVEALED           MessageType = "estimate_revealed"
	PLANNING_RESETED            MessageType = "planning_reseted"
)

type SendMessage struct {
	Type    MessageType `json:"type"`
	Payload interface{} `json:"payload"`
}

type UserJoinedMessage struct {
	User User `json:"user"`
}

type UserDisconnectedMessage struct {
	User User `json:"user"`
}

type ConfirmConnectionMessage struct {
	User           User   `json:"user"`
	ConnectedUsers []User `json:"connectedUsers"`
	Room           Room   `json:"room"`
}

type ConfirmEstimateSubmissionMessage struct {
	Estimate string `json:"estimate"`
}

type UserSubmittedEstimate struct {
	User     User   `json:"user"`
	Estimate string `json:"estimate"`
}

type PlanningResetedMessage struct {
}

func mapRmRoomToRoom(rmRoom *rm.Room) *Room {
	return &Room{
		RoomCode: rmRoom.Code,
		RoomName: rmRoom.Name,
	}
}